### PR TITLE
add div methods on bigint

### PR DESCRIPTION
**Pull request by 6293**
Closes https://github.com/LNP-BP/rust-amplify/issues/90

### CODE REVIEW

File: num/src/error.rs [hunk 1]
```diff
@@ -37,6 +37,21 @@ impl core::fmt::Display for OverflowError {
...
+pub enum DivError {
+    ZeroDiv,
+    Overflow,
+}
+
+impl core::fmt::Display for DivError {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        match *self {
+            DivError::ZeroDiv => write!(f, "division by zero"),
+            DivError::Overflow => write!(f, "division with overflow"),
```
Comment (dr-orlovsky) [inline num/src/error.rs:50]: We normally use `#[derive(Display)]` for that, providing the struct with doc comments and `#[display(doc_comments)]`

File: num/src/error.rs [hunk 2]
```diff
@@ -37,6 +37,21 @@ impl core::fmt::Display for OverflowError {
 #[cfg(feature = "std")]
 impl std::error::Error for OverflowError {}
 
+#[derive(Debug, Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
```
Comment (dr-orlovsky) [inline num/src/error.rs:40]:
```suggestion
#[derive(Debug, Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash, Display, Error)]
#[display(doc_comments)]
```

File: num/src/error.rs [hunk 3]
```diff
@@ -37,6 +37,21 @@ impl core::fmt::Display for OverflowError {
...
+pub enum DivError {
+    ZeroDiv,
```
Comment (dr-orlovsky) [inline num/src/error.rs:42]:
```suggestion
    /// division by zero
    ZeroDiv,
```

File: num/src/error.rs [hunk 4]
```diff
@@ -37,6 +37,21 @@ impl core::fmt::Display for OverflowError {
...
+    Overflow,
```
Comment (dr-orlovsky) [inline num/src/error.rs:43]:
```suggestion
    /// division with overflow
    Overflow,
```

File: num/src/bigint.rs [hunk 5]
```diff
@@ -696,7 +859,7 @@ macro_rules! construct_bigint {
             type Output = $name;
 
             fn div(self, other: T) -> $name {
-                self.div_rem(other.into()).0
+                self.div_rem(other.into()).unwrap().0
```
Comment (dr-orlovsky) [inline num/src/bigint.rs:862]: We never do `unwrap` in non-test code and should use `expect()`. Here, something like
```suggestion
                self.div_rem(other.into()).expect("overflow during big integer division").0
```
will work better

File: num/src/bigint.rs [hunk 6]
```diff
@@ -716,7 +879,7 @@ macro_rules! construct_bigint {
             type Output = $name;
 
             fn rem(self, other: T) -> $name {
-                self.div_rem(other.into()).1
+                self.div_rem(other.into()).unwrap().1
```
Comment (dr-orlovsky) [inline num/src/bigint.rs:882]: Ibid

**Review by dr-orlovsky — CHANGES_REQUESTED** [review]
Two nits, the rest is looking good.

Sorry for the late review, was ill for last week+.

**Comment by 6293** [comment]
Thanks for the review, `unwrap`s are fixed to `expect`. 

> We normally use #[derive(Display)] for that

We have to support `no-std` in ampify-num but `derive` does not support `no-std`. Maybe that is why `OverflowError` in `error.rs` does not use `Display` derive?

**Comment by dr-orlovsky** [comment]
You are right. Sorry, by bad

**Review by dr-orlovsky — APPROVED** [review]
ACK bcb147dfd2398ed61b8c6384829899eb24253320

**Comment by dr-orlovsky** [comment]
Seems like it needs a rebase now

**Comment by 6293** [comment]
rebased

**Review by dr-orlovsky — APPROVED** [review]
ACK a4564c82180808e04f83a810b6cd39ab31b2061d

**Pull request merged**